help/add-a-custom-linkifier.md

```diff
@@ -28,35 +27,133 @@ topic in the message recipient bar that links to the appropriate URL.
 * URL template: `https://github.com/{org}/{repo}/issues/{id}`
 * Original text: `zulip/zulip#2468`
 * Automatically links to: `https://github.com/zulip/zulip/issues/2468`
 
-Linkifier patterns are regular expressions, using the
+{end_tabs}
+
+### Link to a hexadecimal issue or ticket number
+
+The following pattern linkfies a string of hexadecimal digits between 7 and 40
+characters long.
+
+{start_tabs}
+
+* Pattern: `(?P<id>[0-9a-f]{7,40})`
+* URL template: `https://github.com/zulip/zulip/commit/{id}`
+* Original text: `abdc123`
+* Automatically links to: `https://github.com/zulip/zulip/commit/abcd123`
+
+{end_tabs}
+
+## Advanced linkifier patterns
+
+Linkifiers are a flexible system that can be used to construct rules for a wide
+variety of situations. Linkifier patterns are regular expressions, using the
 [re2](https://github.com/google/re2/wiki/Syntax) regular expression
 engine.
 
-If you have any trouble setting these up, please [contact
-us](/help/contact-support) with details on what you're trying to do,
-and we'll be happy to help you out.
+Linkifiers use [RFC 6570](https://www.rfc-editor.org/rfc/rfc6570.html) compliant
+URL templates to describe how links should be generated. These templates support
+several expression types. The default expression type (`{var}`) will URL-encode
+special characters like `/` and `&`; this behavior is desired for the vast
+majority of linkifiers. Fancier URL template expression types can allow you to
+get the exact behavior you want in corner cases like optional URL query
+parameters. For example:
+
+- Use `{+var}` when you want URL delimiter characters to not be URL-encoded.
+
+- Use `{?var}` and `{&var}` when generating URL query parameters.
+
+-  Use `{\#var}` when generating `#` fragments in URLs.
+
+The URL template specification has [brief
+examples](https://www.rfc-editor.org/rfc/rfc6570.html#section-1.2) and [detailed
+examples](https://www.rfc-editor.org/rfc/rfc6570.html#section-3.2) explaining
+the precise behavior of URL templates.
+
+### Linking to documentation pages
+
+This example pattern is a shorthand for linking to pages on Zulip's ReadTheDocs
+site.
+
+{start_tabs}
+
+* Pattern: `RTD/(?P<article>[a-zA-Z0-9_/.#-]+)`
+* URL template: `https://zulip.readthedocs.io/en/latest/{+article}`
+* Original text: `RTD/overview/changelog.html`
+* Automatically links to: `https://zulip.readthedocs.io/en/latest/overview/changelog.html`
+
+{end_tabs}
+
+!!! tip ""
+
+    This pattern uses the `{+var}` expression type. With the
+    default expression type (`{article}`), the `/` between `overview` and
+    `changelog` would incorrectly be URL-encoded
+
+### Linking to Google search results
+
+This example pattern allows linking to Google searches.
+
+{start_tabs}
+
+* Pattern: `google:(?P<query>\w+)?`
+* URL template: `https://google.com/search{?query}`
+* Original text: `google:foo` or `google:`
+* Automatically links to: `https://google.com/search?q=foo` or `https://google.com/search`
```

Comment:
This example has been broken by changing `q` to `query`. As written, it will actually expand to `https://google.com/search?query=foo` which the Google server will not understand.

Reply:
Fixed, thanks.